ShellPkg/Library/UefiShellDriver1CommandsLib/DrvDiag.c

```diff
@@ -199,17 +199,21 @@ DoDiagnostics (
                               );
               if (!EFI_ERROR (Status) && (DriverDiagnostics2 != NULL)) {
                 Language = GetBestLanguageForDriver (DriverDiagnostics2->SupportedLanguages, Lang, FALSE);
-                Found    = TRUE;
-                Status   = DriverDiagnostics2->RunDiagnostics (
-                                                 DriverDiagnostics2,
-                                                 ControllerHandleList[ControllerHandleListLoop],
-                                                 ChildHandleList == NULL ? NULL : ChildHandleList[ChildHandleListLoop],
-                                                 (EFI_DRIVER_DIAGNOSTIC_TYPE)Mode,
-                                                 Language,
-                                                 &ErrorType,
-                                                 &OutBufferSize,
-                                                 &OutBuffer
-                                                 );
+                if (Language == NULL) {
+                  return EFI_NOT_FOUND;

```

Comment:
At the end of the function:
```
  if (DriverHandleList != NULL) {
    FreePool (DriverHandleList);
  }

  if (ControllerHandleList != NULL) {
    FreePool (ControllerHandleList);
  }

  if (ChildHandleList != NULL) {
    FreePool (ChildHandleList);
  }
```
shouldn't we use a goto to free these pointers ?

Reply:
Agreed, switching to a goto.

Reply:
Also freed the other pointers that get allocated and fixed up the rest of the function to use goto.